Enhancement#54

Jayakumar2812 opened the pull request:
## ✅ Enhancement: Add Full Support for `env`, `envOr`, and Array-Based Cheatcodes

This PR introduces complete support for all environment-based cheatcodes (`env*`, `envOr*`, and their array variants), fully aligned with Foundry’s specification.

---

### 🔧 What’s Added

- **Primitive Cheatcodes**:
  - `envInt(string)`
  - `envBytes32(string)`
  - `envAddress(string)`
  - `envBool(string)`
  - `envUint(string)`
  - `envString(string)`
  - `envBytes(string)`

- **Array Variants**:
  - `envInt(string,string)`
  - `envAddress(string,string)`
  - `envBool(string,string)`
  - `envBytes32(string,string)`
  - `envString(string,string)`
  - `envUint(string,string)`
  - `envBytes(string,string)`

- **`envOr` Variants**:
  - `envOr(string, address)`
  - `envOr(string, bool)`
  - `envOr(string, bytes)`
  - `envOr(string, bytes32)`
  - `envOr(string, int256)`
  - `envOr(string, uint256)`
  - `envOr(string, string)`
  - `envOr(string, string, address[])`
  - `envOr(string, string, bool[])`
  - `envOr(string, string, bytes[])`
  - `envOr(string, string, bytes32[])`
  - `envOr(string, string, int256[])`
  - `envOr(string, string, string[])`
  - `envOr(string, string, uint256[])`

- **Existence Check**:
  - `envExists(string)`

---

### 🛠️ Fixes & Changes

- **Corrected Misleading Signatures**:
  - In the previous [pr link](https://github.com/a16z/halmos/commit/081ea7402742ff852119184e80d94ebf1a1d9e6a) I added 
  ```solidity
    0x953C097E: "envBytes(bytes)",
    0x6C42F03F: "envBytes(bytes,bytes)",
  ```
  - The Foundry Book documented the following signatures incorrectly as using `bytes`:
    ```
    function envBytes(bytes calldata key)
    function envBytes(bytes calldata key, bytes calldata delimiter)
    ```
  - But the actual implementation uses `string` parameters:
    [Foundry source → cheatcodes.json](https://github.com/foundry-rs/foundry/blob/b42d5121ba3f03099b1d649ab71f2100a212d3f3/crates/cheatcodes/assets/cheatcodes.json#L4444)

  - Updated function selectors:
    ```solidity
    0x4D7BAF06: "envBytes(string)"
    0xDDC2651B: "envBytes(string,string)"
    ```

- **Organized Codebase**:
  - Moved `dict_of_unsupported_cheatcodes` from `cheatcodes.py` to `utils.py` for better modularity and readability.

---

### 📌 Summary

- Adds complete `env*` and `envOr*` cheatcode support
- Ensures compatibility with Foundry’s cheatcode behavior
- Improves code organization for future scalability

---

Feel free to review and suggest any refinements! ✅

0xkarmacoma commented on src/halmos/sevm.py:
```diff
@@ -2281,6 +2281,7 @@ def handle_insufficient_fund_case(
                     depth=ex.context.depth + 1,
                 )
             )
+            print("Insufficient funds for call:", message)
```
```suggestion
```

0xkarmacoma commented on src/halmos/env.py:
```diff
@@ -0,0 +1,312 @@
+        return self._parse_env_var(key, default)
+
+    def env_int(self, key: str, default: str = None) -> int:
+        val = self._parse_env_var(key)  # raises if missing
```
@Jayakumar2812 Copilot raises a good point, seems like we want this instead
```suggestion
        val = self._parse_env_var(key, default)
```

0xkarmacoma commented on src/halmos/env.py:
```diff
@@ -0,0 +1,312 @@
+
+from dotenv import load_dotenv
+
+load_dotenv()  # Load environment variables from .env file
```
this won't load from `.env_test`, right? Don't we want `.env_test` to be `.env` in `tests/regression` actually?

Jayakumar2812 commented:
* Improved fallback value extraction from env.py functions.
* Removed unnecessary conversion of Solidity input values to comma-separated strings and back to calldata encoding after reading from env.py.(originally done for consistency, but was unnecessary)
* Reverted changes made for testing.
* All test casses are passing as intended.
* To verify env test cases (expect 43 passing, 1 failing as intended):
* 
```
uv run halmos --root tests/regression --match-test env
```
 
 or 
 
 ```
halmos --root tests/regression --match-test env
```
